### ISSUE TITLE
[discussion] busymachines-commons-persistence

### ISSUE THREAD

**Issue by lorandszakacs** [busymachines/busymachines-commons#25]
This is the trickiest one by far, proposal, and thoughts are more than welcome.

I propose the following architecture. You have a module named:
`busymachines-commons-persistence` which contains—maybe—various type classes that can yield "id" for types, and some traits for a "repository" that can allow you to find things by that "id". But the value of this generic repository is disputable.

Then, for each persistence tech we create a separate module which depends on this persistence module, and provides less abstract implementations for the types in the `persistence` module.

Or—alternatively—we craft an abstract `${TechName}Repository` trait only in these specific persistence modules. A "repository" which is better tuned and better suited for the specific data base, thus leaving the topmost `persistence` module very thin (which is not a bad thing).

I think it's needless to mention why it would be a horrible thing for the specific tech-stack persistence modules to depend on each other.

At the end of the day, we should be able to provide the user of this module with the tools to do one simple task (and allowing them to handcraft more complex solution, if they wish), namely:

```
I have a type T. I want to persist this thing, I want to fetch it, and do basic CRUD with it—all the while writing the minimal amount of code.
```

So far, we can potentially implement:
- [ ] `busymachines-commons-persistence-postgres`
- [ ] `busymachines-commons-persistence-mongodb`
- [ ] `busymachines-commons-persistence-elastic`
- [ ] `busymachines-commons-persistence-redis`
- [ ] `busymachines-commons-persistence-kafka`
- etc.

Potential Dependencies (varied per stack, may be transitive or not):
- `busymachines-commons-core`
- `busymachines-commons-json`, issue #23 
- akka-actor
- various drivers for each

**Comment by entity279** [comment]
Yup, agreed that it's hard to write anything here. 

It's hard to abstract over the "relationality" of the persistance techologies. 

With Postgres, I find it hard to justify very clear separations (to the point i don't practice them) in the vein of : module x resposible for CRUD for table T1, module y responsible for CRUD for table T2. It;s much more elegant to perform joins, define & use composable querries.

So we could write some CRUD templates for each of the technologies. For SQL at least it might not be all that useful, but it's start.

Also as remark; with most persistance layers Future.traverse(stuff) {persistanceLayerOp} is a good way to crash/overload it. So the design would need to consider bulk operations as much and as often as possible, otherwise Future.traverse is so easy and convenient to write ... ;)

I'm still unsure if we want to do something

**Comment by lorandszakacs** [comment]
@entity279 I agree with the need for bulk support. And I also agree with the problem of separating CRUD responsibilities for relational data.

So I think that—minimally—our approach should be what you said. Just have handy out of the box CRUD for each technology. With the big mention that it should be used only where you really have to do simple stuff, since it's the most annoying thing to write a lot for simple stuff :(

If you can write 3 lines of code to get a "repository" for simple CRUD, I'd still call that a win, even if you cannot support joins with other "repositories".
